### PR TITLE
Ruby refactoring

### PR DESCRIPTION
**Pull request by MrGoumX**
Hello! During #26, I saw several improvements that can be made (imo) in the codebase, so here is the PR

PS: There are several things that you might not like, If there is something that you don't like we can remove it.

### CODE REVIEW

**Review by smortex — COMMENTED** [review]
Regarding style, it would be really cool to configure rubocop to enforce it as part of CI to have a consistent style on the long run.

I added pointers to relevant docs, but we can probably use the config used in our modules as a sane base for this code:
https://github.com/voxpupuli/modulesync_config/blob/master/moduleroot/.rubocop.yml.erb

One commit per file does not make it easy to follow the changes :sob: Please consider commiing each change (or group of related changes) as single commits with proper commit message to make it easier to review :smile:

File: lib/puppet_metadata/aio.rb [hunk 1]
```diff
@@ -1,21 +1,21 @@
 module PuppetMetadata
-  class AIO 
+  class AIO
     COMPATIBLE = {
-      'AlmaLinux'   => 'RedHat',
```
[the remark this answers is not in the document]
Comment (ekohl) [reply lib/puppet_metadata/aio.rb:4]: I'd be fine with the `key` style.

File: lib/puppet_metadata/aio.rb [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 
-    BUILDS = { 
+    BUILDS = {
       # RPM-based
       'RedHat' => {
         '5' => 5..7,
         '6' => 5..7,
         '7' => 5..7,
-        '8' => 5..7,
+        '8' => 5..7
```
[the remark this answers is not in the document]
Comment (ekohl) [reply lib/puppet_metadata/aio.rb:18]: I'm a strong proponent of the `comma` style (so the current code).

File: spec/metadata_spec.rb [hunk 3]
```diff
@@ -65,15 +65,15 @@
             },
             {
               operatingsystem: 'CentOS',
-              operatingsystemrelease: ['7', '8'],
+              operatingsystemrelease: %w[7 8],
```
[the remark this answers is not in the document]
Comment (ekohl) [reply spec/metadata_spec.rb:68]: No strong preference from me here.

File: lib/puppet_metadata/aio.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrGoumX) [reply lib/puppet_metadata/aio.rb:4]: Restored to original.

File: lib/puppet_metadata/aio.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (MrGoumX) [reply lib/puppet_metadata/aio.rb]: Restored to original and followed this style on some new hashes.

File: spec/metadata_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrGoumX) [reply spec/metadata_spec.rb]: Restored to original.

**Comment by MrGoumX** [comment]
I have changed the order of commits and their names. The changes are:

* Freeze hashes
* Updated the implementation of Beaker and GithubActions
* Use class << self in class methods

File: lib/puppet_metadata/beaker.rb [hunk 4]
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
...
+
+    def self.pidfile_workaround?(pidfile_workaround, os)
+      pidfile_workaround && (!pidfile_workaround.is_a?(Array) || pidfile_workaround.include?(os))
+    end
+
+    def self.setfile_options(options)
+      return '' unless options.any?
+
+      "{#{options.map { |key, value| "#{key}=#{value}" }.join(',')}}"
+    end
```
Comment (ekohl) [inline lib/puppet_metadata/beaker.rb]: Wouldn't it be better to have `build_setfile(name, options)`?

File: lib/puppet_metadata/github_actions.rb [hunk 5]
```diff
@@ -1,5 +1,12 @@
 module PuppetMetadata
   class GithubActions
+    PUPPET_RUBY_VERSIONS = {
```
Comment (ekohl) [inline lib/puppet_metadata/github_actions.rb:3]: I like making this a constant hash, but it's not really a Github Actions thing. It's really an `AIO` thing. It should be defined in `PuppetMetadata::AIO`.

Comment (MrGoumX) [reply lib/puppet_metadata/github_actions.rb]: It has been moved.

File: lib/puppet_metadata/github_actions.rb [hunk 6]
```diff
@@ -52,32 +59,16 @@ def puppet_unit_test_matrix
...
-        '2.7'
-      end
-    end
-
-
     def github_action_test_matrix(use_fqdn: false, pidfile_workaround: false)
-      matrix_include = []
-
-      metadata.operatingsystems.each do |os, releases|
+      metadata.operatingsystems.each_with_object([]) do |(os, releases), arr|
```
[the remark this answers is not in the document]
Comment (MrGoumX) [reply lib/puppet_metadata/github_actions.rb]: I changed the name back to `matrix_include`. If you don't like it we can restore it to the original.

File: lib/puppet_metadata/beaker.rb [hunk 7]
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
...
-        end
-      end
-
-      setfile = name
-      setfile += "{#{options.map { |key, value| "#{key}=#{value}" }.join(',')}}" if options.any?
+      if pidfile_workaround?(pidfile_workaround, os)
+        return if IMAGES_EXCEPTIONS[os]&.include?(release)
 
-      human_name = "#{os} #{release}"
+        options[:image] = IMAGES.dig(os, release) unless IMAGES.dig(os, release).nil?
```
[the remark this answers is not in the document]
Comment (MrGoumX) [reply lib/puppet_metadata/beaker.rb]: I agree this is better.

File: lib/puppet_metadata/beaker.rb [hunk 8]
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
...
-          options[:image] = 'ubuntu:xenial-20191212' if release == '16.04'
-        end
-      end
-
-      setfile = name
-      setfile += "{#{options.map { |key, value| "#{key}=#{value}" }.join(',')}}" if options.any?
+      if pidfile_workaround?(pidfile_workaround, os)
+        return if IMAGES_EXCEPTIONS[os]&.include?(release)
 
-      human_name = "#{os} #{release}"
```
[the remark this answers is not in the document]
Comment (MrGoumX) [reply lib/puppet_metadata/beaker.rb:54]: Restored

File: lib/puppet_metadata/beaker.rb (hunk 4 above)
Comment (MrGoumX) [reply lib/puppet_metadata/beaker.rb]: I updated the implementation. If you still don't like it I'm going to restore it. 

Also I found the comment I deleted, that was my mistake. If you want to me to document any changes just tell me to add it 😛

**Comment by smortex** [comment]
Thanks a lot @MrGoumX for splitting the commits by what they do :+1: 

I fail to spot the change of the static methods (`def self.foo`) to the blocks with what I identify as an "eigenclass" `class << self; def foo`. I remember vaguely using similar things while scratching my head in meta-programing scenarios but here I fail to spot what is the goal you want to achieve?

Apart from this point I don't understand, :100: !

File: lib/puppet_metadata/beaker.rb [hunk 9]
```diff
@@ -4,62 +4,77 @@ module PuppetMetadata
...
-    #   Docker as a Beaker hypervisor. This is to work around Docker's
-    #   limitations.
-    #   When a boolean, it's applied on applicable operating systems. On arrays
-    #   it's applied only when os is included in the provided array.
-    #
-    # @return [nil] If no setfile is available
-    # @return [Array<(String, String)>] The beaker setfile description with a readable name
-    def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround: false)
-      return unless os_supported?(os)
+    IMAGES = {
```
Comment (ekohl) [inline lib/puppet_metadata/beaker.rb:7]: Perhaps we can add a comment here that these images have an older systemd which isn't strict and that's why they do work with `PIDFile`. Previously I considered it an implementation detail but now that they're constants it should be documented. Thinking more about it, perhaps we can also find a better name than `IMAGES`. The most literal one would be `IMAGES_WITH_PIDFILE_COMPATIBLE_SYSTEMD` but perhaps there's a better name. Something like `PIDFILE_COMPATIBLE_IMAGES`

File: lib/puppet_metadata/beaker.rb [hunk 10]
```diff
@@ -4,62 +4,77 @@ module PuppetMetadata
...
+      },
+      'Ubuntu' => {
+        '16.04' => 'ubuntu:xenial-20191212',
+      },
+    }.freeze
 
-      name = "#{os.downcase}#{release.tr('.', '')}-64"
+    # There is no CentOS 8 image that works with PIDFile in systemd
+    # unit files
+    IMAGES_EXCEPTIONS = {
```
Comment (ekohl) [inline lib/puppet_metadata/beaker.rb:18]: Here we can also make the name more explicit. How about `PIDFILE_INCOMPATIBLE_OS_VERSIONS` or just `PIDFILE_INCOMPATIBLE`.

**Review by ekohl — COMMENTED** [review]
Overall I think I'm happy with the changes except a bit of bikeshedding. Let me know what you think.

**Comment by MrGoumX** [comment]
@smortex The only reason I changed these class methods to use `class << self` is to not write `self.method` every time. It's just a personal preference.

File: lib/puppet_metadata/beaker.rb (hunk 10 above)
Comment (MrGoumX) [reply lib/puppet_metadata/beaker.rb]: Changed the variable name!

File: lib/puppet_metadata/beaker.rb (hunk 9 above)
Comment (MrGoumX) [reply lib/puppet_metadata/beaker.rb]: Also changed the variable name and added a comment that hopefully adds some context to the variable.

**Review by ekohl — APPROVED** [review]
Looks good. I see you already prepared for squashing so I'd be happy to merge if you did so.

**Comment by MrGoumX** [comment]
Changes squashed!

**Comment by ekohl** [comment]
Thanks!

**Pull request merged**